extensions/functions_logarithmic.yaml

```diff
@@ -102,5 +102,29 @@ scalar_functions:
             name: "base"
             description: "The logarithm base `b` to use"
         return: fp64
+  -
+    name: "log1p"
+    description: >
+      Natural logarithm (base e) of 1 + the value
 
+      log1p(x) => log(1+x)
+    impls:
+      - args:
+          - name: rounding
+            options: [ TIE_TO_EVEN, TIE_AWAY_FROM_ZERO, TRUNCATE, CEILING, FLOOR ]
+            required: false
+          - name: on_domain_error
+            options: [ NAN, ERROR ]
+            required: false
+          - value: fp32
+        return: fp32
```

Comment:
I realize the surrounding functions don't do this but do these arguments need a `nullability` parameter?

Reply:
Which arguments; as in `- value: fp32`?  And by the same logic, would we also want a lot of the arithmetic functions to have a `nullability` parameter then?  Or did you mean something else?

Reply:
@cpcloud why? The default is `MIRROR` (see table [here](https://substrait.io/expressions/scalar_functions/)), which is correct here.

Reply:
Does this signature handle the case of `NULL` inputs? It looks like it only accepts non-nullable inputs.

Reply:
Yes, because `MIRROR` et al are IMO weird special cases simply because the return type derivation expressions lack the generality to describe them. See https://substrait.io/expressions/scalar_functions/#nullability-handling. `MIRROR` and `DECLARED_OUTPUT` ignore the nullability flag of the arguments and accept any combination of both nullable and non-nullable arguments, and `MIRROR` ignores and overrides the nullability flag of the return type with the boolean or of all argument nullability flags.

Reply:
FWIW, for the validator I'll be implementing a generalization of the type derivation grammar that uses `type??` to specify "either nullable or non-nullable" in a type pattern, and `type?parameter` to bind the nullability to the name `parameter`, so variadics aside, you could describe `MIRROR` and `DECLARED_OUTPUT` more flexibly and explicitly if you want. I want to support that if only to avoid *adding* code for all the special cases, even if Substrait itself won't ever support that syntax.